ambari-server/src/main/java/org/apache/ambari/server/topology/DownloadMpacksTask.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.ambari.server.controller.AmbariServer;
+import org.apache.ambari.server.controller.internal.MpackResourceProvider;
+import org.apache.ambari.server.controller.spi.Request;
+import org.apache.ambari.server.controller.spi.RequestStatus;
+import org.apache.ambari.server.controller.utilities.PropertyHelper;
+import org.apache.ambari.server.state.StackId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Downloads (registers) missing (declared by the blueprint or the cluster template but not present) management packs
+ * during cluster creation.
+ */
+public class DownloadMpacksTask {
+  private static final Logger LOG = LoggerFactory.getLogger(DownloadMpacksTask.class);
+
+  private MpackResourceProvider mpackResourceProvider;
+
+  public DownloadMpacksTask(MpackResourceProvider mpackResourceProvider) {
+    this.mpackResourceProvider = mpackResourceProvider;
+  }
+
+  /**
+   * Registers/downloads all required management packs
+   * @param mpackInstances management pack instances decleared in the blueprint and/or the cluster creation template
+   */
+  public void downloadMissingMpacks(Collection<MpackInstance> mpackInstances) {
+    Set<StackId> missingMpacks = mpackInstances.stream().
+      map(MpackInstance::getStackId).
+      filter(this::isStackMissing).
+      collect(toSet());
+
+    for (MpackInstance mpack: mpackInstances) {
+      if (missingMpacks.contains(mpack.getStackId())) {
+        LOG.info("Registering mpack {}", mpack.getStackId());
+        Request mpackRequest = createRequest(mpack.getUrl());
+        try {
+          RequestStatus.Status status = mpackResourceProvider.createResources(mpackRequest).getStatus();
+          if (!RequestStatus.Status.Complete.equals(status)) {
+            throw new RuntimeException("An error occured while registering mpack. Request status: " + status);
+          }
+        }
+        catch (Exception ex) {
+          throw ex instanceof RuntimeException ?
+            (RuntimeException)ex :
+            new RuntimeException("Error occured while registering mpack: " + mpack.getStackId(), ex);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param mpackUri the uri of the mpack
+   * @return the Request to send to {@link MpackResourceProvider} to to register the mpack
+   */
+  private Request createRequest(String mpackUri) {
+    return PropertyHelper.getCreateRequest(
+      ImmutableSet.of(
+        ImmutableMap.of(
+          MpackResourceProvider.MPACK_URI, mpackUri)),
+      ImmutableMap.of()
+    );
+  }
+
+  /**
+   *
+   * @param stackId the mpack (stack) to check
+   * @return {@code true} if the mpack (stack) is not registered, {@code false} if already registered.
+   */
+  boolean isStackMissing(StackId stackId) {
+    AmbariMetaInfo metaInfo = AmbariServer.getController().getAmbariMetaInfo();
```

Comment:
I would really love if either `AmbariMetaInfo` or `AmbariManagementController` could be passed in as dependency instead of the static lookup.